osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets.Difficulty.Preprocessing;
+using osu.Game.Rulesets.Mods;
+
+namespace osu.Game.Rulesets.Difficulty.Skills
+{
+    public abstract class PreStrainSkill : Skill
+    {
```

Comment:
We can't have a complete copy of all other skills in a new class hierarchy. Consider waiting on https://github.com/ppy/osu/pull/15010

Reply:
Because the preskills don't use necessarily in last calculation(OsuDifficultyCal/PerformanceCal) so the preskills may be confused. And we doesn't have pre calculation though we have current, post calculation. In this situation, If we want to use other skill result, we should put it before original calculation. This is why we should separate it to two types. in structure issue, this may be not good, but something like this requires to calculate note relations. As a result, the roles of two types are very different. it just uses same interface.

I can't find relations between this pr and the pr. Did he really write for calculating it?

Please be radical for changing overall structure. We should try all of radical things.

Reply:
If you mean just changing return types from PreStrainSkill[] to Skill[], it is reasonable.

Reply:
This can happen by passing `Skill`s into other `Skill`s instead. Your `PreStrainSkill` structure provides nothing extra or different to `StrainSkill`, it's a copy-paste of `StrainSkill` + `StrainDecaySkill`.

We already have too many classes. The path forward, as per #15010, is to separate out their defining features (sections, strain decay, etc) into classes which can be used to compose skills. The PR isn't perfect, but the theme is there.

Reply:
resolved